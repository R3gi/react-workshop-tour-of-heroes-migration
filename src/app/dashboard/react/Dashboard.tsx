import React, { FC, useEffect, useState } from 'react'
import { Observable, Subscription } from 'rxjs';
import { Hero } from '../../hero';

export const Dashboard: FC<{ heroes$: Observable<Hero[]> }> = ({ heroes$ }) => {
  const [heroes, setHeroes] = useState<Hero[]>();

  useEffect(() => {
    const subscription = heroes$.subscribe(heroes =>  setHeroes(heroes.slice(1, 5)));

    return () => subscription.unsubscribe();
  }, []);

  return (
    <div><h2>Top Heroes</h2>
      <div className="heroes-menu">
        {heroes?.map(hero => (
        <a
          key={hero.id}
          className='heroes-menu__link'
          href={`/detail/${hero.id}`}>
          {hero.name}
        </a>
        ))}
      </div></div>
  )
}
