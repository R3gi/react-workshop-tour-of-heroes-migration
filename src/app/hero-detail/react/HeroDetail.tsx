import React, { ChangeEvent, FC, useEffect, useState } from 'react'
import { Observable } from 'rxjs';
import { Hero } from '../../hero';

export const HeroDetail: FC<{ getHero$: Observable<Hero>; goBack: () => void; save$: (hero: Hero) => void }> = ({ getHero$, goBack, save$ }) => {
  const [hero, setHero] = useState<Hero>();
  const [inputValue, setInputValue] = useState(hero?.name || '');

  useEffect(() => {
    const subscription = getHero$.subscribe(hero => setHero(hero));

    return () => subscription.unsubscribe();
  }, []);

  if (!hero) {
    return <p>loading...</p>;
  }

  const handleChangeInput = (event: ChangeEvent<HTMLInputElement>) => {
    setInputValue(event.target.value);
  }

  const handleSave = () => {
    if (hero && inputValue) {
      save$({ ...hero, name: inputValue! });
    }
  }

  return (
    <div>
      <h2>{hero.name} Details</h2>
      <div><span>id: </span>{hero.id}</div>
      <div>
        <label htmlFor="hero-name">Hero name: </label>
        <input id="hero-name" placeholder="Hero name" onChange={handleChangeInput} value={inputValue} />
      </div>
      <button type="button" onClick={goBack}>go back</button>
      <button type="button" onClick={handleSave}>save</button>
    </div>
  )
}

