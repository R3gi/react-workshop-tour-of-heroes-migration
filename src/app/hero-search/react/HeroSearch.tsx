import React, { ChangeEvent, FC, useEffect, useState } from 'react'
import { Observable } from 'rxjs';
import { Hero } from '../../hero';

export const HeroSearch: FC<{ getHeroes$?: Observable<Hero[]>; search: (term: string) => void }> = ({ getHeroes$, search }) => {
  const [heroes, setHeroes] = useState<Hero[]>([]);

  const fetchHeroes = () => {
    const subscription = getHeroes$?.subscribe(heroes => setHeroes(heroes));

    return subscription;
  };

  useEffect(() => {
    const subscription = fetchHeroes();

    return () => subscription?.unsubscribe();
  }, []);

  const handleChangeInput = (event: ChangeEvent<HTMLInputElement>) => {
    search(event.target.value);
  }

  return (
    <div id="search-component">
      <label htmlFor="search-box">Hero Search</label>
      <input id="search-box" onChange={handleChangeInput} />

      <ul className="search-result">
        {heroes?.map((hero: Hero) =>
          <li key={hero.id}>
            <a className="search-result__link" href={`/detail/${hero.id}`}>
              {hero.name}
            </a>
          </li>
        )}
      </ul>
    </div>
  )
}
