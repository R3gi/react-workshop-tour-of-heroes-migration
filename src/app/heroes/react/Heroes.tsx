import React, { ChangeEvent, FC, MouseEventHandler, useEffect, useState } from 'react'
import { Hero } from '../../hero';
import { Observable } from 'rxjs';

export const Heroes: FC<{ getHeroes$: Observable<Hero[]>; addHero: (name: string) => void; deleteHero: (hero: Hero) => void  }> = ({ getHeroes$, addHero, deleteHero }) => {
  const [heroes, setHeroes] = useState<Hero[]>();
  const [inputValue, setInputValue] = useState('');
  const fetchHeroes = () =>  getHeroes$.subscribe(heroes => setHeroes(heroes));

  useEffect(() => {
    const subscription = fetchHeroes();

    return () => subscription.unsubscribe();
  }, []);

  const handleChangeInput = (event: ChangeEvent<HTMLInputElement>) => {
    setInputValue(event.target.value);
  }

  const handleAddHero = (name: string) => {
    addHero(name);

    fetchHeroes();
  }

  const handleDeleteHero = (hero: Hero) => {
    deleteHero(hero);

    fetchHeroes();
  }

  return (
    <>
      <h2>My Heroes</h2>

      <div>
        <label htmlFor="new-hero">Hero name: </label>
        <input id="new-hero" onChange={handleChangeInput} value={inputValue}  />
        <button type="button" className="add-button" onClick={() => handleAddHero(inputValue)}>
          Add hero
        </button>
      </div>

      <ul className="heroes">
        {heroes?.map((hero) =>
          <li key={hero.id}>
            <a href={`/detail/${hero.id}`}>
              <span className="badge">{hero.id}</span> {hero.name}
            </a>
            <button type="button" className="delete" title="delete hero" onClick={() => handleDeleteHero(hero)}>x</button>
          </li>
        )}
      </ul>
    </>
  )
};
