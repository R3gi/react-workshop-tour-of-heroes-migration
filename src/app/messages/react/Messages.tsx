import React, { FC, useEffect, useState } from 'react'
import { ReplaySubject } from 'rxjs';

export const Messages: FC<{ clear: () => void, messages: ReplaySubject<string> }> = ({ clear, messages }) => {
  const [allMessages, setMessages] = useState<string[]>([]);

  useEffect(() => {
    let ignore = false;
    messages.subscribe(message => {
      if (!ignore) {
        setMessages((prevMessages) => [...prevMessages, message]);
      }
    });

    return () => { ignore = true }
  }, []);

  const handleClick = () => {
    clear();
    setMessages([]);
  }

  return allMessages.length > 0 && <div><h2>Messages</h2>
    <button type="button"
      className="clear"
      onClick={handleClick}>
      Clear messages</button>
    {allMessages.map((message, index) => {
      return <div key={index}>{message}</div>;
    })}
  </div>;
}
