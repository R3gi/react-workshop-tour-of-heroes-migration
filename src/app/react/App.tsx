import React, { FC } from 'react'
import { MyReactComponent } from './components/MyReactComponent'

export const App: FC<{ id: string }> = ({ id }) => (<div><MyReactComponent id={id} /></div>)
