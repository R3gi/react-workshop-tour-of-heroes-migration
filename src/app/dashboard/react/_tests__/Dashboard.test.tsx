import { render, waitFor } from '@testing-library/react';
import { Dashboard } from "../Dashboard";
import { Observable, from, of } from 'rxjs';
import React from 'react';

// generate react testing library test for Dashboard
describe('<Dashboard />', () => {
  describe('#render', () => {
    it('should render heroes', async () => {
      const { container }  = render(<Dashboard heroes$={of([{ id:1, name:'foo' }])} />)

      await waitFor(() => {
        expect(container).toMatchSnapshot();
      });
    });

    it('should render empty list', async () => {
      const { container }  = render(<Dashboard heroes$={of([])} />)

      await waitFor(() => {
        expect(container).toMatchSnapshot();
      });
    });
   })
 })
