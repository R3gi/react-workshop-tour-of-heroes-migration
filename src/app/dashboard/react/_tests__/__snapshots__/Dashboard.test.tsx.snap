// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> #render should render empty list 1`] = `
<div>
  <div>
    <h2>
      Top Heroes
    </h2>
    <div
      class="heroes-menu"
    />
  </div>
</div>
`;

exports[`<Dashboard /> #render should render heroes 1`] = `
<div>
  <div>
    <h2>
      Top Heroes
    </h2>
    <div
      class="heroes-menu"
    >
      <a
        class="heroes-menu__link"
        href="/detail/1"
      >
        foo
      </a>
    </div>
  </div>
</div>
`;
